### PR TITLE
More efficient indexing

### LINKED ISSUE #56: Inefficiency in indexing data
https://github.com/BirkhoffG/cfnet/blob/6d5c3ec28351de79e575bbded5e67ab1bab48b03/cfnet/datasets.py#L95

You can write something like:
```
 batch_data = self.dataset[self.indices] 
```

### PR DESCRIPTION
**Pull request by Grahams-Uncle**
This fixes #56

### CODE REVIEW

**Comment by Grahams-Uncle** [comment]
fix #56

File: cfnet/datasets.py [hunk 1]
```diff
@@ -157,7 +157,9 @@ def __next__(self):
             if self.shuffle:
                 self.key = next(self.key_seq)
                 self.indices = jax.random.permutation(self.key, self.indices)
-            batch_data = [self.dataset[i] for i in self.indices[: self.batch_size]]
+            #batch_data = [self.dataset[i] for i in self.indices[: self.batch_size]]
+            batch_data = self.dataset[self.indices[: self.batch_size].tolist()]
```
Comment (BirkhoffG) [inline cfnet/datasets.py:161]: why `tolist()` is necessary here?

File: cfnet/datasets.py [hunk 2]
```diff
@@ -157,7 +157,9 @@ def __next__(self):
...
+            batch_data = list(map(list, zip(*batch_data)))
```
Comment (BirkhoffG) [inline cfnet/datasets.py:162]: What is this line doing?

**Review by BirkhoffG — COMMENTED** [review]
Need some clarifications

File: cfnet/datasets.py (hunk 1 above)
Comment (Grahams-Uncle) [reply cfnet/datasets.py]: I have deleted it, I used it to test the syntax.

File: cfnet/datasets.py (hunk 2 above)
Comment (Grahams-Uncle) [reply cfnet/datasets.py:161]: zip and map are used to transpose the data, otherwise, the indexed result is incorrect for training. i.e. [[X_1, X_2, X_3], [Y_1, Y_2, Y_3]] instead of [[X_1,Y_1],[X_2,Y_2],[X_3,Y_3]].

Comment (Grahams-Uncle) [reply cfnet/datasets.py:161]: The indexed result is tuple, so there is no built-in transpose function.

Comment (BirkhoffG) [reply cfnet/datasets.py:161]: What we want is in fact a tuple `([X_1, X_2, X_3], [y_1, y_2, y_3])`.

In fact, most of the time `X` will be more than one dimension, so it should be `([[x_1, x_2, x_3]], [y_1])` if there is only one data in the batch.

**Review by BirkhoffG — CHANGES_REQUESTED** [review]
I think this should be an issue related to the numpy indexing. Previously, we are using numpy for `Dataset`. See https://github.com/google/jax/issues/5417

I provide suggestions to change the code, which works on Python 3.7. Please make changes and test if it works on your computer.

File: nbs/01_datasets.ipynb [hunk 3]
```diff
@@ -252,7 +252,8 @@
     "            if self.shuffle:\n",
```
Comment (BirkhoffG) [inline nbs/01_datasets.ipynb:252]: <p>Change to:</p><pre class="ql-syntax" spellcheck="false">self.X = jnp.asarray(X)
self.y = jnp.asarray(y)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/BirkhoffG/cfnet/pull/62/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (BirkhoffG) [inline nbs/01_datasets.ipynb:252]: <pre class="ql-syntax" spellcheck="false">Line #50.                batch_data = list(map(tuple, zip(*batch_data)))
</pre><p>delete this line</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/BirkhoffG/cfnet/pull/62/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (BirkhoffG) [inline nbs/01_datasets.ipynb:252]: <pre class="ql-syntax" spellcheck="false">Line #57.                return self.collate_fn(batch_data)
</pre><p>Change to <code>return batch_data</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/BirkhoffG/cfnet/pull/62/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

File: cfnet/datasets.py [hunk 4]
```diff
@@ -157,7 +157,8 @@ def __next__(self):
             if self.shuffle:
                 self.key = next(self.key_seq)
                 self.indices = jax.random.permutation(self.key, self.indices)
-            batch_data = [self.dataset[i] for i in self.indices[: self.batch_size]]
+            batch_data = self.dataset[self.indices[: self.batch_size]]
+            batch_data = list(map(tuple, zip(*batch_data)))
```
Comment (BirkhoffG) [inline cfnet/datasets.py:161]: Delete this line

File: nbs/01_datasets.ipynb (hunk 3 above)
Comment (BirkhoffG) [inline nbs/01_datasets.ipynb:252]: <pre class="ql-syntax" spellcheck="false">Line #30.            self.indices: jax.numpy.array = jax.numpy.arange(
</pre><p>change to </p><pre class="ql-syntax" spellcheck="false">self.indices: np.ndarray = np.arange(self.data_len)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/BirkhoffG/cfnet/pull/62/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (BirkhoffG) [inline nbs/01_datasets.ipynb:252]: <pre class="ql-syntax" spellcheck="false">Line #48.                    self.indices = jax.random.permutation(self.key, self.indices)
</pre><p>change to </p><pre class="ql-syntax" spellcheck="false">self.indices = np.random.permutation(self.indices)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/BirkhoffG/cfnet/pull/62/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

File: cfnet/datasets.py [hunk 5]
```diff
@@ -127,20 +127,18 @@ def __init__(
...
         )  # random number sequence
-        self.key_seq.reserve(
-            len(self)
-        )  # generate some random number as key based on the number of batches
         self.key = next(self.key_seq)  # obtain a random key from the sequence
         self.indices: jax.numpy.array = jax.numpy.arange(
             self.data_len
         )  # available indices in the dataset
+        if self.shuffle:
+            self.indices = jax.random.permutation(self.key, self.indices)
```
Comment (BirkhoffG) [inline cfnet/datasets.py:141]: Use `np.random.permutation(self.indices)`

File: cfnet/datasets.py [hunk 6]
```diff
@@ -127,20 +127,18 @@ def __init__(
...
 
         self.data_len: int = len(dataset)  # Length of the dataset
         self.key_seq: hk.PRNGSequence = hk.PRNGSequence(
             self.seed
         )  # random number sequence
-        self.key_seq.reserve(
-            len(self)
-        )  # generate some random number as key based on the number of batches
         self.key = next(self.key_seq)  # obtain a random key from the sequence
         self.indices: jax.numpy.array = jax.numpy.arange(
```
Comment (BirkhoffG) [inline cfnet/datasets.py:137]: Use `np.arange(...)`

File: cfnet/datasets.py [hunk 7]
```diff
@@ -154,17 +152,14 @@ def __len__(self):
...
         if self.pose <= self.data_len:
-            if self.shuffle:
-                self.key = next(self.key_seq)
-                self.indices = jax.random.permutation(self.key, self.indices)
-            batch_data = [self.dataset[i] for i in self.indices[: self.batch_size]]
+            batch_data = self.dataset[self.indices[: self.batch_size]]
             self.indices = self.indices[self.batch_size :]
             if self.drop_last and len(self.indices) < self.batch_size:
                 self.pose = 0
                 self.indices = jax.numpy.arange(self.data_len)
```
Comment (BirkhoffG) [inline cfnet/datasets.py:159]: Use `np.arange(...)`

**Review by BirkhoffG — CHANGES_REQUESTED** [review]
Change these:
- This current code seems only shuffle once the **entire time**. What we want is to shuffle once at the start of **EVERY EPOCH**.
- Change `self.indices` from `jax.numpy` to original numpy `np`.

**Review by BirkhoffG — APPROVED** [review]
lgtm
